internal/staticanalysis/obfuscation/parsing/babel-parser.js

```diff
@@ -0,0 +1,297 @@
+#!/usr/bin/env node
+/* eslint no-var: 0 */
+
+const parser = require("@babel/parser");
+const fs = require("fs");
+
+function locationString(node) {
+    return (node.loc != null) ? `[${node.loc.start.line},${node.loc.start.column}]` : "[]"
+}
+
+const parseOutputLines = []
+
+function logJSON(type, subtype, name, pos, array, extra = null) {
+    const extraValue = (extra !== null) ? `${JSON.stringify(extra)}` : "{}"
+    const arrayValue = (array !== null) ? array : false
+    const json = `{"type":"${type}","subtype":"${subtype}","data":${JSON.stringify(name)},"pos":${pos},` +
+        `"array":${arrayValue}, "extra":${extraValue}}`
+    parseOutputLines.push(json)
+}
+function logIdentifierJSON(subtype, name, pos, extra = null) {
+    logJSON("Identifier", subtype, name, pos, null, extra)
+}
+
+function logLiteralJSON(subtype, value, pos, inArray, extra = null) {
+    logJSON("Literal", subtype, value, pos, inArray, extra)
+}
+
+function logParameters(node) {
+    const n = node
+    for (let i = 0; i < n.params.length; i++) {
+        p = n.params[i]
+        switch (p.type) {
+            // simple function parameter name:
+            // function f(x)
+            case "Identifier":
+                logIdentifierJSON("Parameter", p.name, locationString(p))
+                break
+            // parameter with default value:
+            // function f(x = 3)
+            case "AssignmentPattern":
+                if (p.left.type === "Identifier") {
+                    logIdentifierJSON("Parameter", p.left.name, locationString(p))
+                } else {
+                    // This is too hard...
+                    walkAst(p.left)
+                }
+                walkAst(p.right)
+                break
+        }
+    }
+
+}
+
+function multiWalkAst(nodeList, isInArray = false) {
+    if (nodeList !== null) {
+        for (let i = 0; i < nodeList.length; i++) {
+            walkAst(nodeList[i], isInArray)
+        }
+    }
+}
+
+function walkAst(startNode, isInArray = false, printDebug = false) {
```

Comment:
Have you looked at https://babeljs.io/docs/en/babel-traverse ?

This might make the traversal a lot easier because you don't have to encode knowledge of the AST structure here.